### PR TITLE
did a bit of refactoring of handleRedisCommand path

### PR DESCRIPTION
**Pull request by sjpotter**
based on review that the functions were being treated as a void, did a bit of a refactor to try and clean it up

### CODE REVIEW

File: src/redisraft.c [hunk 1]
```diff
@@ -38,7 +38,7 @@ void *__dso_handle;
 #define VALID_NODE_ID(x)    ((x) > 0)
 
 static void redirectCommand(RedisRaftCtx *rr, RedisModuleCtx *ctx, Node *leader);
-static RRStatus handleNonLeaderCommand(RedisRaftCtx *rr, RedisModuleCtx *ctx,  Node *leader, RaftRedisCommandArray *cmds);
+static void handleNonLeaderCommand(RedisRaftCtx *rr, RedisModuleCtx *ctx,  Node *leader, RaftRedisCommandArray *cmds);
```
Comment (tezc) [inline src/redisraft.c:41]:
```suggestion
static void handleNonLeaderCommand(RedisRaftCtx *rr, RedisModuleCtx *ctx, Node *leader, RaftRedisCommandArray *cmds);
```

File: src/redisraft.c [hunk 2]
```diff
@@ -591,76 +591,79 @@ static bool handleInterceptedCommands(RedisRaftCtx *rr,
...
-            if (term != rr->snapshot_info.last_applied_term) {
-                replyClusterDown(ctx);
-                return RR_ERROR;
-            }
-
+/* Handle the special case of read-only commands here: if quorum reads
+ * are enabled schedule the request to be processed when we have a guarantee
+ * we're still a leader. Otherwise, just process the reads.
+ *
+ * Normally we can expect a single command in the request, unless it is a
```
Comment (tezc) [inline src/redisraft.c:598]: This comment belongs to where we call `CommandSpecGetAggregateFlags()`

File: src/redisraft.c [hunk 3]
```diff
@@ -591,76 +591,79 @@ static bool handleInterceptedCommands(RedisRaftCtx *rr,
...
+ * MULTI/EXEC transaction in which case all queued commands are handled at
+ * once.
+ */
+static void handleRedisCommandAppendReadOnly(RedisRaftCtx *rr, RedisModuleCtx *ctx,
+                                             RaftRedisCommandArray *cmds) {/* read only command, quorum or not */
```
Comment (tezc) [inline src/redisraft.c:603]:
```suggestion
static void handleRedisCommandAppendReadOnly(RedisRaftCtx *rr, RedisModuleCtx *ctx,
                                             RaftRedisCommandArray *cmds)
{
```

File: src/redisraft.c [hunk 4]
```diff
@@ -591,76 +591,79 @@ static bool handleInterceptedCommands(RedisRaftCtx *rr,
...
+        RaftReq *req = RaftReqInit(ctx, RR_REDISCOMMAND);
+        RaftRedisCommandArrayMove(&req->r.redis.cmds, cmds);
+        raft_queue_read_request(rr->raft, handleReadOnlyCommand, req);
+    } else {
+        /* Wait until the new leader applies an entry from the current term.
+         * Otherwise, we might process a request before replaying logs.
+         * The state machine will be in an older state. Reading from it
+         * might look like going backward in time. */
+        raft_term_t term = raft_get_current_term(rr->raft);
+        if (term == rr->snapshot_info.last_applied_term) {
```
Comment (tezc) [inline src/redisraft.c:615]: I think this makes code less readable. Common pattern is having an if check for error case and returning inside it. So, you keep success path in the main flow. Previous version of this code is easier to understand.

File: src/redisraft.c [hunk 5]
```diff
@@ -591,76 +591,79 @@ static bool handleInterceptedCommands(RedisRaftCtx *rr,
...
             RaftExecuteCommandArray(rr, ctx, ctx, cmds);
+        } else {
+            replyClusterDown(ctx);
         }
-        return RR_OK;
     }
+}
 
+static void handleRedisCommandAppendLog(RedisRaftCtx *rr, RedisModuleCtx *ctx, RaftRedisCommandArray *cmds) {
```
Comment (tezc) [inline src/redisraft.c:623]:
```suggestion
static void handleRedisCommandAppendLog(RedisRaftCtx *rr, RedisModuleCtx *ctx, RaftRedisCommandArray *cmds)
{
```

File: src/redisraft.c [hunk 6]
```diff
@@ -591,76 +591,79 @@ static bool handleInterceptedCommands(RedisRaftCtx *rr,
...
-        return RR_ERROR;
+    } else {
+        raft_entry_release(entry);
     }
+}
 
-    raft_entry_release(entry);
-    return RR_OK;
+static void handleRedisCommandAppend(RedisRaftCtx *rr,
+                                      RedisModuleCtx *ctx, RaftRedisCommandArray *cmds)
```
Comment (tezc) [inline src/redisraft.c:644]:
```suggestion
                                     RedisModuleCtx *ctx, RaftRedisCommandArray *cmds)
```

File: src/redisraft.c [hunk 7]
```diff
@@ -591,76 +591,79 @@ static bool handleInterceptedCommands(RedisRaftCtx *rr,
...
+     * joining or loading data.
+     */
+    if (checkRaftState(rr, ctx) == RR_ERROR) {
+        return;
+    }
+
+    if (!raft_is_leader(rr->raft)) {
+        Node *leader = getLeaderNodeOrReply(rr, ctx);
+
+        return handleNonLeaderCommand(rr, ctx, leader, cmds);
```
Comment (tezc) [inline src/redisraft.c:658]: A bit weird to return with a void function

File: src/redisraft.c (hunk 4 above)
Comment (sjpotter) [reply src/redisraft.c:615]: ok, changed back, was trying to avoid extra returns

File: src/redisraft.c [hunk 8]
```diff
@@ -591,76 +591,79 @@ static bool handleInterceptedCommands(RedisRaftCtx *rr,
...
+
+    /* Check that we're part of a bootstrapped cluster and not in the middle of
+     * joining or loading data.
+     */
+    if (checkRaftState(rr, ctx) == RR_ERROR) {
+        return;
+    }
+
+    if (!raft_is_leader(rr->raft)) {
+        Node *leader = getLeaderNodeOrReply(rr, ctx);
```
Comment (tezc) [inline src/redisraft.c:656]: We are not doing anything with it here, we can call this inside `handleNonLeaderCommand()`.

File: src/redisraft.c (hunk 7 above)
Comment (sjpotter) [reply src/redisraft.c:658]: the other option was

```
handleNonLeaderCommand(rr, ctx, leader, cmds);
return;
```

is that better?

Comment (sjpotter) [reply src/redisraft.c:658]: combined with the next if/else block to avoid need for return

File: src/redisraft.c [hunk 9]
```diff
@@ -591,76 +591,80 @@ static bool handleInterceptedCommands(RedisRaftCtx *rr,
...
+
     int e = raft_recv_entry(rr->raft, entry, &req->r.redis.response);
     if (e != 0) {
         replyRaftError(ctx, e);
         entryDetachRaftReq(rr, entry);
         raft_entry_release(entry);
         RaftReqFree(req);
-        return RR_ERROR;
+    } else {
+        raft_entry_release(entry);
```
Comment (tezc) [inline src/redisraft.c:639]: As both paths are calling `raft_entry_release(entry)`, we can call this unconditionally but I think I find more readable when we separate error paths from the success path and keep success path in the main flow. (Not in an "if check" if possible).  Previously, when we get an error, we were calling bunch of functions and we were returning from the error case. I find it easier to understand.

Comment (sjpotter) [reply src/redisraft.c:639]: ok, reverted else and put void return back in

File: src/redisraft.c [hunk 10]
```diff
@@ -697,42 +701,36 @@ static void redirectCommand(RedisRaftCtx *rr, RedisModuleCtx *ctx, Node *leader)
...
-static RRStatus handleNonLeaderCommand(RedisRaftCtx *rr, RedisModuleCtx *ctx, Node *leader, RaftRedisCommandArray *cmds)
+static void handleNonLeaderCommand(RedisRaftCtx *rr, RedisModuleCtx *ctx, Node *leader, RaftRedisCommandArray *cmds)
 {
     RedisModule_Assert(!raft_is_leader(rr->raft));
     RedisModule_Assert(cmds);
 
-    /* reply ASK for ASKING state commands */
-    if (cmds->asking) {
+
+    if (!leader) {
```
Comment (tezc) [inline src/redisraft.c:711]: Going back to same discussion :) I'd prefer previous version. I find it harder to understand when there are bunch of if checks. It's common to have "fallthrough" if checks for error cases/conditions and I find it easier to read:

```c
if (!leader) {
    .....
    return; 
}

if (rr->config->follower_proxy) {
    /* forward commands to the leader in follower_proxy state */
   .......
    return;
}

if (asking) {
    ......
    return;
}

redirect();
return;
```

File: src/redisraft.c [hunk 11]
```diff
@@ -591,76 +591,80 @@ static bool handleInterceptedCommands(RedisRaftCtx *rr,
...
+        return handleNonLeaderCommand(rr, ctx, leader, cmds);
+    }
+
+    if (cmd_flags & CMD_SPEC_UNSUPPORTED) {
+        RedisModule_ReplyWithError(ctx, "ERR not supported by RedisRaft");
+    } else if (cmd_flags & CMD_SPEC_READONLY && !(cmd_flags & CMD_SPEC_WRITE)) {
+        handleRedisCommandAppendReadOnly(rr, ctx, cmds);
+    } else {
+        handleRedisCommandAppendLog(rr, ctx, cmds);
+    }
```
Comment (tezc) [inline src/redisraft.c:667]:
```suggestion
    if (cmd_flags & CMD_SPEC_UNSUPPORTED) {
        RedisModule_ReplyWithError(ctx, "ERR not supported by RedisRaft");
        return;
    } 
    
    if (cmd_flags & CMD_SPEC_READONLY && !(cmd_flags & CMD_SPEC_WRITE)) {
        handleRedisCommandAppendReadOnly(rr, ctx, cmds);
    } else {
        handleRedisCommandAppendLog(rr, ctx, cmds);
    }
```

For example, I find this one easier to understand.

File: src/redisraft.c (hunk 10 above)
Comment (sjpotter) [reply src/redisraft.c:711]: ok. redid

File: src/redisraft.c (hunk 11 above)
Comment (sjpotter) [reply src/redisraft.c]: ok, done

File: src/redisraft.c (hunk 7 above)
Comment (sjpotter) [reply src/redisraft.c]: reverted above comment's change.

**Review by tezc — APPROVED** [review]

**Pull request merged**